File: svc/backtests/clients/go/backtests.go
// Generate code for mock
//go:generate go run go.uber.org/mock/mockgen@v0.2.0 -source=backtests.go -destination=mock.gen.go -package client

package client

import (
	"context"
	"time"

	client "github.com/lerenn/cryptellation/pkg/client"
	"github.com/lerenn/cryptellation/pkg/models/account"
	"github.com/lerenn/cryptellation/pkg/models/event"
	"github.com/lerenn/cryptellation/pkg/models/order"
	"github.com/lerenn/cryptellation/svc/backtests/pkg/backtest"
	"github.com/lerenn/cryptellation/svc/candlesticks/pkg/period"

	"github.com/google/uuid"
)

type Client interface {
	Advance(ctx context.Context, backtestID uuid.UUID) error
	Create(ctx context.Context, payload BacktestCreationPayload) (uuid.UUID, error)
	CreateOrder(ctx context.Context, payload client.OrderCreationPayload) error
	Get(ctx context.Context, backtestID uuid.UUID) (backtest.Backtest, error)
	GetAccounts(ctx context.Context, backtestID uuid.UUID) (map[string]account.Account, error)
	Subscribe(ctx context.Context, backtestID uuid.UUID, exchange, pair string) error
	ListenEvents(ctx context.Context, backtestID uuid.UUID) (<-chan event.Event, error)
	List(ctx context.Context) ([]backtest.Backtest, error)
	ListOrders(ctx context.Context, backtestID uuid.UUID) ([]order.Order, error)

	ServiceInfo(ctx context.Context) (client.ServiceInfo, error)
	Close(ctx context.Context)
}

type BacktestCreationPayload struct {
	Accounts    map[string]account.Account
	StartTime   time.Time
	EndTime     *time.Time
	Mode        *backtest.Mode
	PricePeriod *period.Symbol
}

File: clients/python/cryptellation/backtests.py

from .utils import cryptellation
from .displayer import Displayer

class Backtest(object):
    '''Backtest class'''

    def _load_data(self):
        return cryptellation(['backtests', 'get', self.id])
    
    def _load_candlesticks(self):
        return cryptellation(['candlesticks', 'read',
            '--exchange', self.data['prices_subscriptions'][0]['exchange'],
            '--pair', self.data['prices_subscriptions'][0]['pair'],
            '--period', self.data['parameters']['price_period'],
            '--start', self.data['parameters']['start_time'],
            '--end', self.data['parameters']['end_time']
            ])

    def __init__(self, id):
        self.id = id
        self.data = self._load_data()

    def analyze(self):
        '''Analyze the backtest'''
        self.candlesticks = self._load_candlesticks()
        self.orders = self.list_orders()
        Displayer(self.candlesticks, self.orders).display()

    def list_orders(self):
        '''List all orders'''
        return cryptellation(['backtests', 'orders', 'list', self.id])

def list():
    '''List all backtests'''
    return cryptellation(['backtests', 'list'])

def last():
    '''Get the last backtest'''
    return Backtest(list()[-1]['id'])